modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/TransactionValidator.scala

```diff
@@ -0,0 +1,207 @@
+package org.tessellation.dag.transaction
+
+import cats.data.ValidatedNel
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.traverse._
+import cats.syntax.validated._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Balance
+import org.tessellation.schema.transaction.{Transaction, TransactionReference}
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.key.ops.PublicKeyOps
+import org.tessellation.security.signature.Signed
+import org.tessellation.security.signature.signature.SignatureProof
+
+import eu.timepit.refined.auto._
+import io.estatico.newtype.ops._
+
+trait TransactionValidator[F[_]] {
+  import TransactionValidator._
+
+  implicit val F: Async[F]
+  implicit val kryoSerializer: KryoSerializer[F]
+  implicit val securityProvider: SecurityProvider[F]
+
+  protected def getBalance(address: Address): F[Balance]
+
+  protected def getLastAcceptedTransactionRef(address: Address): F[TransactionReference]
+
+  def validate(
```

Comment:
you should be able to do either:
```
def validate(tx) = implementation
def validate(txs) => txs.traverse(validate)
```
or
```
def validate(tx) = validate(Seq(tx))
def validate(txs) = implementation
```

I'd opt for the second. Definitely you don't need the implementation twice

Reply:
The first one won't work as we need to validate balances of all the transactions together. The second one should work.